ledger/store/src/program/finalize.rs

```diff
@@ -553,196 +406,134 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
-            None => Ok(None),
+        // Retrieve the key-values for the mapping.
+        match self.key_value_id_map().get_map_speculative(&(program_id, mapping_name))? {
+            Some(key_value_ids) => Ok(key_value_ids),
+            None => Ok(Vec::new()),
         }
     }
 
     /// Returns the confirmed value for the given `program ID`, `mapping name`, and `key`.
     fn get_value_confirmed(
         &self,
-        program_id: &ProgramID<N>,
-        mapping_name: &Identifier<N>,
+        program_id: ProgramID<N>,
+        mapping_name: Identifier<N>,
         key: &Plaintext<N>,
     ) -> Result<Option<Value<N>>> {
-        // Retrieve the key ID.
-        match self.get_key_id_confirmed(program_id, mapping_name, key)? {
-            // Retrieve the value.
-            Some(key_id) => self.get_value_from_key_id_confirmed(&key_id),
+        match self.key_value_id_map().get_value_confirmed(&(program_id, mapping_name), key)? {
+            Some(value) => Ok(Some(cow_to_cloned!(value))),
             None => Ok(None),
         }
     }
 
     /// Returns the speculative value for the given `program ID`, `mapping name`, and `key`.
     fn get_value_speculative(
         &self,
-        program_id: &ProgramID<N>,
-        mapping_name: &Identifier<N>,
+        program_id: ProgramID<N>,
+        mapping_name: Identifier<N>,
         key: &Plaintext<N>,
     ) -> Result<Option<Value<N>>> {
-        // Retrieve the key ID.
-        match self.get_key_id_speculative(program_id, mapping_name, key)? {
-            // Retrieve the value.
-            Some(key_id) => self.get_value_from_key_id_speculative(&key_id),
-            None => Ok(None),
-        }
-    }
-
-    /// Returns the confirmed value for the given `key ID`.
-    fn get_value_from_key_id_confirmed(&self, key_id: &Field<N>) -> Result<Option<Value<N>>> {
-        match self.value_map().get_confirmed(key_id)? {
+        match self.key_value_id_map().get_value_speculative(&(program_id, mapping_name), key)? {
             Some(value) => Ok(Some(cow_to_cloned!(value))),
             None => Ok(None),
         }
     }
 
-    /// Returns the speculative value for the given `key ID`.
-    fn get_value_from_key_id_speculative(&self, key_id: &Field<N>) -> Result<Option<Value<N>>> {
-        match self.value_map().get_speculative(key_id)? {
-            Some(value) => Ok(Some(cow_to_cloned!(value))),
-            None => Ok(None),
-        }
+    /// Returns the confirmed checksum of the finalize storage.
+    fn get_checksum_confirmed(&self) -> Result<Field<N>> {
+        // Compute all mapping checksums.
+        let preimage: std::collections::BTreeMap<_, _> = self
+            .key_value_id_map()
+            .iter_confirmed()
+            .map(|(m, k, v)| {
+                let m = cow_to_copied!(m);
+                let k = cow_to_cloned!(k);
+                let v = cow_to_cloned!(v);
+
+                let mut preimage = Vec::new();
```

Comment:
this vector can be bound before the iterations so that its allocation can be reused in all of them; it just needs to be cleared once `entry_checksum` is ready